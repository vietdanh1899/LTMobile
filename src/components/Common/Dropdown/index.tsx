/* eslint-disable no-nested-ternary */
import React, { PureComponent } from 'react';
import {
  Text,
  View,
  TouchableOpacity,
  TouchableWithoutFeedback,
  ScrollView,
  StyleSheet,
  Dimensions,
} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { withTheme, ThemeProps } from 'react-native-elements';

const { width, height } = Dimensions.get('window');

const styles = StyleSheet.create({
  textInput: {
    fontWeight: 'bold',
  },
  containerInput: {
    // borderWidth: 1,
    paddingHorizontal: 15,
    height: 50,
    borderRadius: 10,
    flexDirection: 'row',
    justifyContent: 'space-between',
    // backgroundColor: '#FFFFFF',
    // textAlign: 'center',
    // alignContent: 'center',
    alignItems: 'center',
    // zIndex: 1000,
    // elevation: 1000,
  },
  isSelected: {
    // /borderWidth: 1,
    padding: 15,
    borderRadius: 10,
    flexDirection: 'row',
    justifyContent: 'space-between',
    // backgroundColor: '#315DF7',
    // top: -10,
  },
  isNotSelected: {
    // /borderWidth: 1,
    padding: 15,
    borderRadius: 10,
    flexDirection: 'row',
    justifyContent: 'space-between',
    backgroundColor: '#FFFFFF',
    // top: -10,
  },
  // textOption: {
  //   fontSize:
  // }
  rotateIcon: {
    transform: [{ rotate: '180deg' }],
  },

  containerDropdownStyles: {
    overflow: 'scroll',
    backgroundColor: '#FFFFFF',
    borderRadius: 10,
    position: 'absolute',
  },
  infoViewStyles: {
    width: '90%',
    justifyContent: 'center',
  },
});

interface Props {
  defaultValue?: number | string;
  onChange?: any;
  containerStyle?: any;
  dropdownStyles?: any;
  activeColor?: string;
  activeTextColor?: string;
  data: any;
  inActiveTextColor?: string;
  inActiveColor?: string;
  placeholderColor?: string;
  iconComponent?: any;
  iconSize?: number;
  iconColor?: string;
  placeholder?: string;
  theme?: any;
  // inActiveComponent?: any;
  // activeComponent?: any;
}
interface State {
  rotateIcon: boolean;
  value: number | string;
}

class Dropdown extends PureComponent<Props, State> {
  constructor(props: any) {
    super(props);
    const { defaultValue, data } = this.props;
    let valueSeclected = data[0].value;
    if (defaultValue) {
      valueSeclected = defaultValue;
    }
    this.state = {
      rotateIcon: false,
      value: valueSeclected,
    };
  }

  setValue = (value: any) => () => {
    const { onChange } = this.props;
    if (onChange) {
      onChange(value);
    }
    this.setState({
      value,
      rotateIcon: false,
    });
  };

  render() {
    const {
      containerStyle,
      dropdownStyles,
      activeColor,
      activeTextColor,
      data,
      placeholderColor,
      inActiveColor,
      inActiveTextColor,
      iconComponent,
      iconSize,
      iconColor,
      placeholder,
      theme,
    } = this.props;
    const { rotateIcon, value } = this.state;
    let labelSelected = placeholder || 'Select item';
    // let checkWidth: any;

    if (value) {
      data.forEach((item: any) => {
        if (value === item.value) {
          labelSelected = item.label;
        }
      });
    } else {
      labelSelected = placeholder || 'Select item';
    }

    const textStyles = StyleSheet.flatten([
      styles.textInput,
      { color: placeholderColor || theme.Dropdown.placeholderColor },
    ]);

    const isNotSelectedStyles = StyleSheet.flatten([
      styles.isNotSelected,
      { backgroundColor: inActiveColor || theme.colors.bgColor },
    ]);

    return (
      <View>
        <View style={{ zIndex: 10 }}>
          <TouchableWithoutFeedback
            onPress={() => this.setState({ rotateIcon: !rotateIcon })}
          >
            <View
              style={[
                styles.containerInput,
                { backgroundColor: theme.Dropdown.backgroundColor },
                containerStyle,
              ]}
            >
              <View style={styles.infoViewStyles}>
                <Text numberOfLines={1} style={[{ fontSize: 16 }, textStyles]}>
                  {labelSelected}
                </Text>
              </View>
              <View
                style={[
                  rotateIcon && styles.rotateIcon,
                  { justifyContent: 'center' },
                ]}
              >
                {iconComponent || (
                  <Icon
                    name="chevron-down"
                    size={iconSize || 24}
                    color={iconColor || theme.Dropdown.iconColor}
                  />
                )}
              </View>
            </View>
          </TouchableWithoutFeedback>
          {rotateIcon ? (
            <View
              style={[
                styles.containerDropdownStyles,
                { backgroundColor: theme.colors.bgColor },
                dropdownStyles,
              ]}
            >
              <ScrollView>
                {data.map((item: any) => (
                  <TouchableOpacity
                    key={item.value}
                    onPress={this.setValue(item.value)}
                  >
                    <View
                      style={
                        value === item.value
                          ? [
                            styles.isSelected,
                            activeColor
                              ? { backgroundColor: activeColor }
                              : {
                                backgroundColor: theme.Dropdown.activeColor,
                              },
                          ]
                          : isNotSelectedStyles
                      }
                    >
                      <Text
                        style={
                          value === item.value
                            ? activeTextColor
                              ? { color: activeTextColor }
                              : { color: theme.Dropdown.activeTextColor }
                            : {
                              color:
                                  inActiveTextColor || theme.Dropdown.textColor,
                            }
                        }
                      >
                        {item.label}
                      </Text>
                    </View>
                  </TouchableOpacity>
                ))}
              </ScrollView>
            </View>
          ) : null}
        </View>
        {rotateIcon ? (
          <TouchableWithoutFeedback
            onPress={() => this.setState({ rotateIcon: !rotateIcon })}
          >
            <View
              style={{
                flex: 1,
                position: 'absolute',
                left: 0,
                top: 0,
                // opacity: 0.5,
                // backgroundColor: 'black',
                backgroundColor: 'transparent',
                width,
                height,
                justifyContent: 'center',
              }}
            >
              {/* <ActivityIndicator /> */}
            </View>
          </TouchableWithoutFeedback>
        ) : null}
      </View>
    );
  }
}
// export default Dropdown;
export default withTheme(
  (Dropdown as unknown) as React.ComponentType<Props & ThemeProps<any>>,
);
