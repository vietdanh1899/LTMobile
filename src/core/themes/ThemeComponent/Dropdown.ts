import { darkComponentColor, lightComponentColor } from './Common/CommonProps';
import { Color } from './Common/Color';

/**
 * Header
 */
export const darkDropdown = {
  backgroundColor: darkComponentColor.bgColorSecondary,
  placeholderColor: Color.white,
  iconColor: Color.white,
  activeColor: darkComponentColor.bgColorSecondary,
  activeTextColor: Color.white,
  textColor: Color.white,
};
export const lightDropdown = {
  backgroundColor: lightComponentColor.bgColorSecondary,
  placeholderColor: Color.black,
  iconColor: Color.black,
  activeColor: lightComponentColor.bgColorSecondary,
  activeTextColor: Color.black,
  textColor: Color.black,
};
