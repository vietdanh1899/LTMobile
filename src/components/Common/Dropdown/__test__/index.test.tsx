import React from 'react';
import {
  EnumComponent, basicTest, getComponent,
} from '@utils/testHelper';
import Dropdown from '..';

const dataDropdown = [
  {
    label: 'Dự án Complex Quy Nhon',
    value: 1,
  },
  {
    label: 'Partner',
    value: 2,
  },
  {
    label: 'Partner1',
    value: 3,
  },
  {
    label: 'Lorem ipsum dolor sit',
    value: 4,
  },
  {
    label: 'architecto voluptatum. Neque?',
    value: 5,
  },
  {
    label: 'Lorem Lorem, ipsum',
    value: 6,
  },
  {
    label: 'Dự án đất nền Đà Nẵng',
    value: 7,
  },
];
const target = <Dropdown data={dataDropdown} />;
const type = EnumComponent.HOC;
describe('Dropdown Component', () => {
  basicTest(target, type, 1, false, false);
  it('should apply all custom props', () => {
    const initialTest = getComponent(target, { containerStyle: { borderWidth: 1 } }, type).props();
    expect(initialTest.children[0].props.children[0].props.children.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          borderWidth: 1,
        }),
      ]),
    );
  });
});
