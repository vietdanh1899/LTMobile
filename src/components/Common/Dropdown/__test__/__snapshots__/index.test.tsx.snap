// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component should render without issues 1`] = `
<View>
  <View
    style={
      Object {
        "zIndex": 10,
      }
    }
  >
    <TouchableWithoutFeedback
      onPress={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 10,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-between",
              "paddingHorizontal": 15,
            },
            Object {
              "backgroundColor": "#dee2e6",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "justifyContent": "center",
              "width": "90%",
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "#121212",
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Dự án Complex Quy Nhon
          </Text>
        </View>
        <View
          style={
            Array [
              false,
              Object {
                "justifyContent": "center",
              },
            ]
          }
        >
          <Icon
            allowFontScaling={false}
            color="#121212"
            name="chevron-down"
            size={24}
          />
        </View>
      </View>
    </TouchableWithoutFeedback>
  </View>
</View>
`;
